### PR TITLE
log based replication scheam changes

### PR DESCRIPTION
**Pull request by kfir-payne**
Signed-off-by: Kfir Payne <kfirpayne@gmail.com>

### Explain the changes
1. add new fields in the schema to support log based replication info
2. validate logs bucket exist
3. change bucket api in accordance

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1. 

- [ ] Doc added/updated
- [ ] Tests added

### CODE REVIEW

File: src/server/system_services/replication_store.js [hunk 1]
```diff
@@ -84,7 +84,6 @@ class ReplicationStore {
         return ans;
     }
 
-
```
Comment (dannyzaken) [inline src/server/system_services/replication_store.js:87]: remove change

File: src/api/bucket_api.js [hunk 2]
```diff
@@ -1522,5 +1528,19 @@ module.exports = {
                 }
             }
         },
+
+        log_based_replication_info: {
```
Comment (dannyzaken) [inline src/api/bucket_api.js:1532]: maybe move `log_based_replication_info` into `replication policy` definition? they seem to always go together. 

Also, I think that `log_based_replication_info` is a bit of a mouthful. maybe rename it to `log_replication_info` or something?

File: src/server/system_services/bucket_server.js [hunk 3]
```diff
@@ -1973,6 +1973,14 @@ function validate_replication(req) {
             throw new RpcError('INVALID_REPLICATION_POLICY', `Rule ${rule_id} destination bucket is invalid`);
         }
 
+        if (rule.log_based_replication_info) {
```
Comment (dannyzaken) [inline src/server/system_services/bucket_server.js:1976]: `log_based_replication_info` is not part of `rule` anymore. should be moved to out of the for loop

File: src/server/system_services/bucket_server.js [hunk 4]
```diff
@@ -1977,6 +1977,25 @@ function validate_replication(req) {
...
+            req.system.buckets_by_name[req.rpc_params.replication_policy.log_replication_info.logs_location.logs_bucket.unwrap()];
+
+        // validate logs_bucket exist
+        if (!logs_bucket) {
+            throw new RpcError('INVALID_LOG_REPLICATION_INFO', `logs bucket is invalid`);
+        }
+
+        const logs_bucket_info = get_bucket_info({ bucket: logs_bucket });
+        if (logs_bucket_info.namespace && logs_bucket.namespace.write_resource.resource.connection.endpoint_type !== 'AWS') {
+            throw new RpcError('INVALID_LOG_REPLICATION_INFO', `logs bucket is not compatible with log based replication`);
```
Comment (romayalon) [inline src/server/system_services/bucket_server.js:1992]: we need to add here to the error message that logs_bucket should be a namespace bucket on AWS

File: src/server/system_services/bucket_server.js [hunk 5]
```diff
@@ -1977,6 +1977,25 @@ function validate_replication(req) {
...
+        }
+        if (logs_bucket_info.mode !== 'OPTIMAL') {
```
Comment (romayalon) [inline src/server/system_services/bucket_server.js:1994]: if we are throwing on the operator a persistent error and rejecting, maybe for this case we shouldn't

File: src/server/system_services/bucket_server.js [hunk 6]
```diff
@@ -1998,13 +2017,41 @@ function validate_replication(req) {
...
+    const validated_replication = { rules: validated_replication_rules };
+
+    if (req.rpc_params.replication_policy.log_replication_info) {
+        let validated_log_based_replication;
+        const logs_bucket = req.system.buckets_by_name &&
+            req.system.buckets_by_name[req.rpc_params.replication_policy.log_replication_info.logs_location.logs_bucket.unwrap()];
+        if (!logs_bucket) {
+            const source_bucket = find_bucket(req);
+            validated_log_based_replication = {
+                log_replication_info: { logs_location: { logs_bucket: source_bucket._id } }
```
Comment (romayalon) [inline src/server/system_services/bucket_server.js:2037]: why doing it?

File: src/server/system_services/bucket_server.js [hunk 7]
```diff
@@ -1998,13 +2017,41 @@ function validate_replication(req) {
...
+            };
+        } else {
+            validated_log_based_replication = {
+                log_replication_info: { logs_location: { logs_bucket: logs_bucket._id } }
+            };
+        }
+        if (req.rpc_params.replication_policy.log_replication_info.logs_location.subpath !== "") {
```
Comment (romayalon) [inline src/server/system_services/bucket_server.js:2044]: why all this code is needed? 
I don't see why putting '/' for no subpath and '' for marker, keeping it empty won't work?

File: src/server/system_services/bucket_server.js (hunk 5 above)
Comment (kfir-payne) [reply src/server/system_services/bucket_server.js:1994]: worth to make a discussion about it, @dannyzaken WDYT?

File: src/server/system_services/bucket_server.js (hunk 6 above)
Comment (kfir-payne) [reply src/server/system_services/bucket_server.js:2037]: its one of the default behaviour we agreed on, if logs_bucket is not provided assume the logs are in the source bucket

Comment (romayalon) [reply src/server/system_services/bucket_server.js:2037]: so add it as a comment please here

File: src/server/system_services/bucket_server.js (hunk 7 above)
Comment (kfir-payne) [reply src/server/system_services/bucket_server.js:2044]: putting '/' on the subpath it's just a default value if non provided, search the logs at the top of the bucket.
log marker can stay empty, planed to initiate it on the BG worker, and in the meanwhile put a placeholder in order to insert to the DB.

File: src/server/system_services/bucket_server.js (hunk 6 above)
Comment (kfir-payne) [reply src/server/system_services/bucket_server.js:2037]: will do

File: src/server/system_services/bucket_server.js (hunk 7 above)
Comment (romayalon) [reply src/server/system_services/bucket_server.js:2047]: Without '/' searching the logs will work? why do we need in the DB an empty marker? you can just check if it exists or not

Comment (kfir-payne) [reply src/server/system_services/bucket_server.js:2047]: we could leave them empty and then when needed check existence and proceed as follows, I think that placing the default values now will save us time on always checking if they exist.

File: src/server/system_services/bucket_server.js (hunk 5 above)
Comment (dannyzaken) [reply src/server/system_services/bucket_server.js:1995]: I think we can skip this check for now

File: src/server/system_services/schemas/replication_configuration_schema.js [hunk 8]
```diff
@@ -39,6 +39,25 @@ module.exports = {
...
                 }
+            },
+        },
+        log_replication_info: {
+            type: 'object',
+            required: ['logs_location'],
+            properties: {
+                logs_location: {
+                    type: 'object',
+                    required: ['logs_bucket', 'subpath'],
```
Comment (dannyzaken) [inline src/server/system_services/schemas/replication_configuration_schema.js:50]: I think `subpath` shouldn't be required. maybe rename it to prefix?

File: src/server/system_services/schemas/replication_configuration_schema.js [hunk 9]
```diff
@@ -39,6 +39,29 @@ module.exports = {
...
+        aws_log_replication_info: {
+            type: 'object',
+            required: ['logs_location'],
+            properties: {
+                logs_location: {
+                    type: 'object',
+                    required: ['logs_bucket'],
+                    properties: {
+                        logs_bucket: {
+                            objectid: true // bucket id
```
Comment (dannyzaken) [inline src/server/system_services/schemas/replication_configuration_schema.js:53]: didn't we decide to access the logs will through the S3 bucket directly? 
of course, the user will have to provide a bucket that is accessible by the same account

Comment (kfir-payne) [reply src/server/system_services/schemas/replication_configuration_schema.js:53]: if remember correctly we decided that the logs bucket will also be a namespace bucket in Noobaa

File: src/server/system_services/bucket_server.js [hunk 10]
```diff
@@ -1932,10 +1932,16 @@ async function get_bucket_replication(req) {
...
-    const bucket_names_replication = _.map(replication, rule =>
+    const bucket_names_replication = _.map(replication.rules, rule =>
         ({ ...rule, destination_bucket: system_store.data.get_by_id(rule.destination_bucket).name }));
 
-    return bucket_names_replication;
+    let { logs_location } = replication.aws_log_replication_info ? replication.aws_log_replication_info : undefined;
+    const res = {
+        rules: bucket_names_replication,
+        log_replication_info: logs_location
+    };
```
Comment (dannyzaken) [inline src/server/system_services/bucket_server.js:1942]: if `aws_log_replication_info` is undefined, I think your code will fail.
you can use optional chaining, which is supported in node 14, or you can just use a simple `if`

with optional chaining
```suggestion
    const res = {
        rules: bucket_names_replication
        log_replication_info: replication?.aws_log_replication_info?.logs_location
    }
```


or:
```suggestion
    const res = {
        rules: bucket_names_replication
    }
    if (replication.aws_log_replication_info) {
        res.log_replication_info = aws_log_replication_info.logs_location
    }
```

**Review by dannyzaken — APPROVED** [review]

**Pull request merged**